File: demo-python-service/patterns/transaction/idempotency.py
import hashlib
import json
import time
import threading
from typing import Any, Callable, Dict, Optional

class IdempotencyService:
    def __init__(self, ttl: int = 300):  # 5 minutes default TTL
        self.cache: Dict[str, Dict[str, Any]] = {}
        self.ttl = ttl
        self.lock = threading.Lock()

    def generate_key(self, *args, **kwargs) -> str:
        data = {'args': args, 'kwargs': kwargs}
        json_str = json.dumps(data, sort_keys=True, default=str)
        return hashlib.sha256(json_str.encode()).hexdigest()

    def execute_idempotent(self, key: str, operation: Callable) -> Any:
        with self.lock:
            if key in self.cache:
                cached_entry = self.cache[key]
                if time.time() - cached_entry['timestamp'] < self.ttl:
                    return cached_entry['result']
                else:
                    del self.cache[key]

        result = operation()
        
        with self.lock:
            self.cache[key] = {
                'result': result,
                'timestamp': time.time()
            }
        
        return result

    def invalidate(self, key: str) -> bool:
        with self.lock:
            return self.cache.pop(key, None) is not None

    def get_stats(self) -> Dict[str, Any]:
        with self.lock:
            return {
                'cached_entries': len(self.cache),
                'ttl_seconds': self.ttl
            }

File: demo-node-service/patterns/transaction/idempotency.js
const crypto = require('crypto');

class IdempotencyService {
    constructor(ttl = 300000) { // 5 minutes default TTL
        this.cache = new Map();
        this.ttl = ttl;
    }

    generateKey(...args) {
        const data = JSON.stringify(args, null, 0);
        return crypto.createHash('sha256').update(data).digest('hex');
    }

    async executeIdempotent(key, operation) {
        if (this.cache.has(key)) {
            const cached = this.cache.get(key);
            if (Date.now() - cached.timestamp < this.ttl) {
                return cached.result;
            }
            this.cache.delete(key);
        }

        const result = await operation();
        this.cache.set(key, {
            result,
            timestamp: Date.now()
        });

        // Auto cleanup after TTL
        setTimeout(() => this.cache.delete(key), this.ttl);
        
        return result;
    }

    invalidate(key) {
        return this.cache.delete(key);
    }

    getStats() {
        return {
            cachedEntries: this.cache.size,
            ttlMs: this.ttl
        };
    }
}

module.exports = { IdempotencyService };